File: problems/Design an Ordered Stream/orderedStream.js
/**
 * @param {number} n
 */
var OrderedStream = function(n) {
    this.store = Array(n);
    this.visited = new Set();
    this.idx = 0;
};

/**
 * @param {number} idKey
 * @param {string} value
 * @return {string[]}
 */
OrderedStream.prototype.insert = function(idKey, value) {
    const result = [];

    this.store[idKey - 1] = value;
    this.visited.add(idKey - 1);

    while (this.visited.has(this.idx)) {
        result.push(this.store[this.idx++]);
    }

    return result;
};

/**
 * Your OrderedStream object will be instantiated and called as such:
 * var obj = new OrderedStream(n)
 * var param_1 = obj.insert(idKey,value)
 */

File: problems/Design an Ordered Stream/orderedStream.java
class OrderedStream {

    public OrderedStream(int n) {
        store = new String[1 + n];
        Arrays.fill(store, "%");
        idx = 1;
    }

    public List<String> insert(int idKey, String value) {
        List<String> result = new ArrayList<>();

        store[idKey] = value;

        while (store.length > idx && !"%".equals(store[idx])) {
            result.add(store[idx++]);
        }

        return result;
    }

    String[] store;
    int idx;
}

/**
 * Your OrderedStream object will be instantiated and called as such:
 * OrderedStream obj = new OrderedStream(n);
 * List<String> param_1 = obj.insert(idKey,value);
 */
